apps/showcase/src/components/training/code-editor-view/code-editor-view.component.ts

```diff
@@ -0,0 +1,199 @@
+
+  /**
+   * Service to load files and run commands in the application instance of the webcontainer.
+   */
+  public readonly webContainerService = inject(WebContainerService);
+  /**
+   * File tree loaded in the project folder within the web container instance.
+   */
+  public tree$ = this.webContainerService.monacoTree$;
+  /**
+   * @see {FormBuilder}
+   */
+  private readonly formBuilder = inject(FormBuilder);
+
+  /**
+   * Form with the selected file and its content which can be edited in the Monaco Editor
+   */
+  public form: FormGroup<{
+    code: FormControl<string | null>;
+    file: FormControl<string | null>;
+  }> = this.formBuilder.group({
+      code: '',
+      file: this.project?.startingFile || ''
+    });
+  /**
+   * Configuration for the Monaco Editor
+   */
+  public editorOptions$ = this.form.controls.file.valueChanges.pipe(
+    startWith(this.project?.startingFile || ''),
+    filter((filePath): filePath is string => !!filePath),
+    map((filePath) => ({
+      theme: 'vs-dark',
+      language: editorOptionsLanguage[filePath.split('.').pop() || 'ts'] || editorOptionsLanguage.ts,
+      readOnly: (this.editorMode === 'readonly'),
+      automaticLayout: true,
+      scrollBeyondLastLine: false
+    }))
+  );
+
+  constructor() {
+    this.form.controls.code.valueChanges.pipe(
+      distinctUntilChanged(),
+      skip(1),
+      debounceTime(300),
+      filter((text): text is string => !!text),
+      takeUntilDestroyed()
+    ).subscribe((text: string) => {
+      if (!this.project) {
+        // eslint-disable-next-line no-console
+        console.error('No project found');
+        return;
+      }
+      const path = `${this.project.cwd}/${this.form.controls.file.value}`;
+      // eslint-disable-next-line no-console
+      console.log('Writing file', path);
+      void this.webContainerService.writeFile(`${this.project.cwd}/${this.form.controls.file.value}`, text);
+    });
+    this.form.controls.file.valueChanges.pipe(
+      // Force to re-emit if the tree changes
+      combineLatestWith([
+        this.tree$.pipe(filter((tree): tree is MonacoTreeElement[] => tree.length > 0))
+      ]),
+      switchMap(([path]) => {
+        if (!this.project) {
+          return of('');
+        }
+        return from(this.webContainerService.readFile(`${this.project.cwd}/${path}`));
+      }),
+      takeUntilDestroyed()
+    ).subscribe((content) => this.form.controls.code.setValue(content));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public async onClickFile(filePath: string) {
+    if (!this.project) {
+      return;
+    }
+    const path = `${this.project.cwd}/${filePath}`;
+    if (this.project && await this.webContainerService.isFile(path)) {
+      this.form.controls.file.setValue(filePath);
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public ngOnChanges(changes: SimpleChanges) {
+    if ('project' in changes) {
+      if (this.project && changes.project.currentValue !== changes.project.previousValue) {
```

Comment:
`changes.project.currentValue !== changes.project.previousValue`  might be redundant